Prototype 1.2

SMoragon opened the pull request.

SMoragon committed:
First prototype iteration.
Now, the full workflow option just returns the query that will need to be executed, waiting for fixing some workflow bugs to execute it instead.
Also, some simple options, such as simple inference, align and inference workflow and so on probably will be deleted (design decissions are to be made).
Last but not least, this will not work if MEvoLib is not installed (TODO: include it in the user guide/README).

SMoragon committed:
Logic changes.
Now, the interface allows the user to select an input file from a list of valid given ones. They can also add an output file format in all stages, and the conversion between files will be checked (if input is not fasta and the conversion tuple is not supported) to prevent users from launching a failing workflow.

JAlvarezJarreta committed:
cleanup unnecessary files

SMoragon committed:
Bugs and typos fixed.
Also, full query built has been slightly changed to adapt and ease the query to the incoming development and Nextflow integration.

SMoragon committed:
Merge branch 'Prototype1' of https://github.com/nachorequeno/mevolib-ui into Prototype1

SMoragon committed:
Fatinataly changes applied

SMoragon committed:
Many changes:

-> Old pull request changes applied, including SECRET_KEY management (now, it is stored locally through python-dotenv) and non-necessary files deleted.
-> pyproject.toml added, to ensure users use and acknowledge of all of the required dependecies, apart from development ones, the repository location and a few more options.
-> Migrations have been reset to be only 1 file.
-> The most crucial feature added: now, the Nextflow workflows are launched asynchronously. For doing it, Celery and Redis have been integrated into the project, and this commit is a small example of it's future integration with the full workflow. Also, a folder name is given to each task. This all will ensure concurrency when this server is deployed.

To do:

-> Integrate recent changes in mevolib-workflow and MEvoLib repositories.
-> Clean up all of the features that will not be used, such as the simple alignment.
-> Integrate Celery and the full MEvoLIb workflow, to accomplish the original task of this GUI.

SMoragon committed:
Forgotten files

SMoragon committed:
Typos fixed

SMoragon committed:
Minor errors fixed

SMoragon commented:
**Minor errors** commit also includes buildFullQuery method refactor, as the integration with MEvoLib and MEvoLib workflow has started.
Also, the Celery/Redis management of the full workflow has been made, still not tested because of mevolib-workflow having few trouble.

SMoragon committed:
MEvoLib files parameters (in/out formats) synced up, typos and bugs fixed and new cluster_tool parameter added.
As a TODO, the error workflow log has to be managed, still figuring out how to do that...

SMoragon committed:
MEvoLib integration has finished.
Also, the app now downloads a zip named after the output file name the user put.
Last but not least, some bugs and typos have been fixed, and a couple more validations regarding launching an empty workflow have been added.

SMoragon committed:
Pyproject errors fixed

JAlvarezJarreta commented on MEvoLibGUI/pyproject.toml:
```diff
@@ -0,0 +1,75 @@
+[project.optional-dependencies]
+dev = [
+    "black",
+    "pylint",
```
```suggestion
    "mypy",
    "pylint",
```

JAlvarezJarreta commented on MEvoLibGUI/pyproject.toml:
```diff
@@ -0,0 +1,75 @@
+    "version",
+]
+requires-python = ">= 3.10.6"
+description = "Graphical User Interface for a DNA analysis library (MEvoLIb)"
```
```suggestion
description = "Graphical User Interface for a DNA analysis library (MEvoLib)"
```

JAlvarezJarreta commented on MEvoLibGUI/pyproject.toml:
```diff
@@ -0,0 +1,75 @@
+description = "Graphical User Interface for a DNA analysis library (MEvoLIb)"
+readme = "README.md"
+authors = [
+    {name = "Sergio Moragón Jiménez", email = "smoragonj@gmail.com"}
```
Based on personal experience, I would suggest you to drop accents as it may cause issues in some countries.

JAlvarezJarreta commented on MEvoLibGUI/pyproject.toml:
```diff
@@ -0,0 +1,75 @@
+    "MEvoLib",
+]
+[tool.setuptools]
+py-modules = []
+
```
```suggestion
```
If empty, remove :smile_cat:

SMoragon replied:
Is needed

SMoragon committed:
Update MEvoLibGUI/pyproject.toml

Co-authored-by: J. Alvarez-Jarreta <jalvarez@ebi.ac.uk>

SMoragon committed:
Many changes:

- Old simple workflows deleted, including HTML, scripts, urls, views and settings references, among others.
- Workflow optimized: noew, if all goes fine, the user will receive a repor to alongside the folder. On the oter hand, if there is any error, a log with the information will be hsown, so that they can know what went wrong.
- Error modal added, so that the alert that was before is no longer needed.
- README modified to include all of the installation and execution process, aprat from a pair of common bugs and how to solve them (still under development, as a solution for cloning MEvoLib and mevolib-workflow is being tested yet)

SMoragon committed:
Merge branch 'Prototype1' of https://github.com/nachorequeno/mevolib-ui into Prototype1

JAlvarezJarreta commented on MEvoLibGUI/MEvoLibGUI/tasks.py:
```diff
@@ -0,0 +1,37 @@
+                    flog.write(f"{line}\n")
+
+        raise RuntimeError("An error has ocurred")
+        
+    
+
+    
+  
```
```suggestion
```

JAlvarezJarreta commented on MEvoLibGUI/MEvoLibGUI/tasks.py:
```diff
@@ -0,0 +1,37 @@
+from pathlib import Path
+import nextflow, os, re
+
+from MEvoLibGUI.settings import (
+    WORKFLOW_OUTPUT as wf_output,
+)
```
```suggestion
from MEvoLibGUI.settings import WORKFLOW_OUTPUT as wf_output
```

JAlvarezJarreta commented on MEvoLibGUI/MEvoLibGUI/tasks.py:
```diff
@@ -0,0 +1,37 @@
+from celery import shared_task
+from pathlib import Path
+import nextflow, os, re
```
```suggestion
from pathlib import Path

from celery import shared_task
import nextflow, os, re
```

JAlvarezJarreta commented on MEvoLibGUI/MEvoLibGUI/tasks.py:
```diff
@@ -0,0 +1,37 @@
+    execution = nextflow.run(workflow_path, params=parameters, report = report_dir)
+    output_dir = Path(parameters["output_dir"])
+    
+    if not Path.exists(output_dir.absolute()):
+        output_dir.mkdir(parents=True, exist_ok=True)
+        log_path = output_dir / "nextflow.log"
+        
+        with log_path.open("w") as flog:
+            flog.write(f"{execution.command}\n\n")
+            for line in execution.log.splitlines():
+                if workflow_error.search(line) or task_error.search(line):
+                    flog.write(f"{line}\n")
+
+        raise RuntimeError("An error has ocurred")
```
If you still want to create the log file when only a few jobs failed, you may need to change this code:
```suggestion
    output_dir.mkdir(parents=True, exist_ok=True)
    log_path = output_dir / "nextflow.log"
    with log_path.open("w") as flog:
        flog.write(f"{execution.command}\n\n")
        for line in execution.log.splitlines():
            if workflow_error.search(line) or task_error.search(line):
                flog.write(f"{line}\n")
    if execution.return_code == 1:
        raise RuntimeError("An error has ocurred")
```
Whilst I'm not 100% sure this code will work, it should based on the documentation of nextflowpy.

SMoragon replied:
I will test it

JAlvarezJarreta commented on MEvoLibGUI/MEvoLibGUI/tasks.py:
```diff
@@ -0,0 +1,37 @@
+    """Orders a Nextflow workflow to start, so that it can be executed asynchronously while the user keeps on
+    doing other tasks."""
+
+    workflow_error = re.compile(r"^.+ ERROR .+$")
+    task_error = re.compile(r"^.+ terminated with an error .+$")
```
I would suggest to make these two constants.

SMoragon replied:
Gotcha

JAlvarezJarreta commented on MEvoLibGUI/MEvoLibGUI/tasks.py:
```diff
@@ -0,0 +1,37 @@
+
+    workflow_error = re.compile(r"^.+ ERROR .+$")
+    task_error = re.compile(r"^.+ terminated with an error .+$")
+    
+    report_dir = Path(parameters["output_dir"]).joinpath("execution_report.html")
+    
+    execution = nextflow.run(workflow_path, params=parameters, report = report_dir)
+    output_dir = Path(parameters["output_dir"])
+    
```
```suggestion
    output_dir = Path(parameters["output_dir"])
    report_file = output_dir / "execution_report.html"
    execution = nextflow.run(workflow_path, params=parameters, report=report_file)
    # Create log file with relevant information about the workflow execution
```

JAlvarezJarreta reviewed (changes requested):
Haven't finished reviewing all the code yet, but I have a couple of requests/question besides my comments:
- Can you please add the license header to all the files, please?
- Why are all the mevolib-workflow files again here? Shouldn't it be a dependency for your repo? As installation instructions, since it cannot be added as a Python dependency.

SMoragon commented:
> Haven't finished reviewing all the code yet, but I have a couple of requests/question besides my comments:
> 
> * Can you please add the license header to all the files, please?
> * Why are all the mevolib-workflow files again here? Shouldn't it be a dependency for your repo? As installation instructions, since it cannot be added as a Python dependency.

I will add them the sooner I can.
Also, sorry for mevolib-workflow files being included, I forgot to include it to the .gitignore files.

SMoragon committed:
License headers added, style changes applied, and error report modified, so that the nextflow.log file is always among the user's ZIP, just in case some tasks have failed during the execution but did not stop it (as the error strategy is to ignore them).
